### PR TITLE
CVE-2015-1234 and CVE-2016-5149

### PR DESCRIPTION
**Pull request by gldias**
Research, curate, and record findings on CVE-2015-1234 and CVE-2016-5149

### CODE REVIEW

File: cves/CVE-2015-1234.yml [hunk 1]
```diff
@@ -47,16 +58,18 @@ reviews:
...
+  note:
 vccs:
-- :commit:
-  :note: 
+- commit: 939e7367d5d984efe1f402c7e5135503beaebfae
+  note: |
+    They were adding a wrapper to support BOOL types. There is a specific line that
+    was particularly responsible for the CVE: GLsizei num_values = result->GetNumResults();
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
```
Comment (angbamfield) [inline cves/CVE-2015-1234.yml]: +2 upvotes

File: cves/CVE-2015-1234.yml [hunk 2]
```diff
@@ -77,9 +90,11 @@ unit_tested:
 
     For the "fix" answer below, check if the fix for the vulnerability involves
     adding or improving an automated test to ensure this doesn't happen again.
-  answer: 
-  code: 
-  fix: 
+  answer: |
+    There are unit tests for the code in the file in question, but it seems like
+    no tests were modified or added because of the vulnerability.
```
Comment (angbamfield) [inline cves/CVE-2015-1234.yml]: Do you think there is a reason why they decided not to add tests to verify the vulnerability fix?

File: cves/CVE-2015-1234.yml [hunk 3]
```diff
@@ -132,7 +160,10 @@ major_events:
 
     The event doesn't need to be directly related to this vulnerability, rather,
     we want to capture what the development team was dealing with at the time.
-  answer: 
+  answer: |
+    No specific event was found, but it seems like the file in question was being worked on by
+    multiple people across some big companies (e.g. Intel, Samsung, Nvidia). Too many developers
+    could have lead to multiple issues.
```
Comment (angbamfield) [inline cves/CVE-2015-1234.yml]: Mention some basic issues with the "too many developers" thought (ie. miscommunication leading to other vulnerabilities), other than that good explanation.

File: cves/CVE-2015-1234.yml [hunk 4]
```diff
@@ -196,6 +236,27 @@ mistakes:
     they have written there. Are they doing those? Does the fix look proper?
 
     Use those questions to inspire your answer. Don't feel obligated to answer
-    every one. Write a thoughtful entry here that those ing the software
+    every one. Write a thoughtful entry here that those in the software
     engineering industry would find interesting.
-  answer:
```
Comment (angbamfield) [inline cves/CVE-2015-1234.yml]: no changes needed here just wanted to say this was a really thoughtful explanation on the mistakes that were made !

File: cves/CVE-2016-5149.yml [hunk 5]
```diff
@@ -50,20 +64,27 @@ reviews:
...
+  note: |
+    While this did not introduce the vulnerability by itself, it contains a line of code in script_context.cc
+    that uses an extension URL that is potentially dangerous, and allowed for the vulnerability to be
+    exploited.
+- commit: f91cb89ccc9f4621d6b8d8928f70788d212def9b
+  note: |
+    This commit failed to validate URLs, extensions, and/or iframes, which is why the vulnerability
+    was possible in the first place.
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: +2 upvotes

File: cves/CVE-2016-5149.yml [hunk 6]
```diff
@@ -84,9 +105,9 @@ unit_tested:
 
     For the "fix" answer below, check if the fix for the vulnerability involves
     adding or improving an automated test to ensure this doesn't happen again.
-  answer: 
-  code: 
-  fix: 
+  answer: The fix commit indicates that tests were present and new tests are added to prevent this vulnerability.
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: What did the old/new tests check for? Any specific edge cases ?

File: cves/CVE-2016-5149.yml [hunk 7]
```diff
@@ -125,12 +146,18 @@ interesting_commits:
...
-  - commit: 
-    note: 
+  - commit: 62b5911d512673e84cf5b2c63e9b20ec5ad2c3b8
+    note: |
+      This commit reverts 4 patches that had previously been merged. There were a lot of attempts to fix
+      bugs in the Extensions sybsystem, and many of these attempts were reverted at least once. Seems like
+      code was being moved around a lot, which may have caused some confusion/diffusion among developers.
+  - commit: ee4adef0b70f13cb80a886ef639fc910c0b5ce6f
+    note: |
+      This is a very large commit started enabling new API features for the Extension subsystem. This indicates that
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: So the developers were working on a new feature so this vulnerability ended up being hidden, is that what you are trying to say ?

File: cves/CVE-2016-5149.yml [hunk 8]
```diff
@@ -50,20 +64,27 @@ reviews:
...
   CVE-2011-3092.yml). Fixes and VCCs follow the same format.
 fixes:
-- :commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
-  :note: ''
-- :commit: 79b64c3e741cc9c6afbb23885945831a45c6baa5
-  :note: ''
-- :commit: 4b8ca0a9fbbe79fa5a7819917e5a31d237f0f499
-  :note: ''
+- commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
+  note: 'Original fix'
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: Original Fix by who? what was the fix?

File: cves/CVE-2016-5149.yml [hunk 9]
```diff
@@ -102,20 +123,20 @@ discovered:
 
     If there is no evidence as to how this vulnerability was found, then you may
     leave the entries blank except for "answer". Write down where you looked in "answer".
-  answer: 
-  date: 
-  automated: 
-  google: 
-  contest: 
+  answer: Reported by user with the email justi...@gmail.com (https://bugs.chromium.org/p/chromium/issues/detail?id=573131).
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: how did this person discover this? Manual, fuzzer, testing, etc...?

File: cves/CVE-2016-5149.yml [hunk 10]
```diff
@@ -205,4 +242,15 @@ mistakes:
...
+    by far, how long it took to react to it. According to the bug report it took over
+    6 months to resolve this bug, which is an unbelievable amount of time for a
+    high-severity and high-priority bug like this one.
+
+    The fix involved some simple validation and some new tests, which may indicate an oversight in the
+    original VCC, although this vulnerability would have been difficult to predict
+    since it relied on a race condition. However, it was a severe lack of communication
+    that lead to this vulnerability living as long as it did. There were long periods of time
+    where there was little to no effort being put into resolving this bug, which exposed
+    this application to really dangerous attacks.
```
Comment (angbamfield) [inline cves/CVE-2016-5149.yml]: did any dangerous attacks actually happen or like what types of dangerous attacks could've happened?

Comment (gldias) [reply cves/CVE-2016-5149.yml]: This vulnerability could, for example, allow an attacker to access extension APIs that they would not normally have access to, which can help them spoof an extension and interact with other extensions in a malicious way.

File: cves/CVE-2016-5149.yml [hunk 11]
```diff
@@ -205,4 +246,15 @@ mistakes:
...
     every one. Write a thoughtful entry here that those ing the software
     engineering industry would find interesting.
-  answer: 
+  answer: |
+    I think that the biggest issue that came up during this vulnerability's life span was,
+    by far, how long it took to react to it. According to the bug report it took over
+    6 months to resolve this bug, which is an unbelievable amount of time for a
+    high-severity and high-priority bug like this one.
+
+    The fix involved some simple validation and some new tests, which may indicate an oversight in the
```
Comment (ConR24) [inline cves/CVE-2016-5149.yml]: Do you think that this vulnerability would count as a design mistake, that's what it sounds like in your second paragraph

File: cves/CVE-2016-5149.yml [hunk 12]
```diff
@@ -192,7 +228,12 @@ lessons:
     note: 
   complex_inputs:
     applies: 
-    note: 
+    note:
+  concurrency_is_a_risk:
+    applies: true
+    note: |
+      One of the issues that made this vulnerability possible was a race condition between HTML document
```
Comment (ConR24) [inline cves/CVE-2016-5149.yml]: I feel like this could be better if it was fleshed out more, but it's fine in its current state.

File: cves/CVE-2015-1234.yml [hunk 13]
```diff
@@ -118,12 +137,23 @@ interesting_commits:
...
+      They were experiencing multiple minor bugs, so they preemptively added logging functionality
+      to multiple files, including the file that introduced the vulnerability. While this is a
+      safe and effective way to diagnose future issues, it seems like they were logging a bitt too
+      much, just by looking at the amount of log commands in each file within this commit.
+  - commit: 0140c18b7e341dd2c4e81f9323e830c69574cda1
+    note: |
+      They seemed to be doing a lot of work with GPU buffers, which is heavily related to the
+      vulnerability. Even still, the vulnerability was only found years after the VCC was introduced.
+      This leads me to believe that, even though the vulnerability was critical, it was mostly
+      hidden behind another weakness that was not really being tested. While this vulnerability
```
Comment (ConR24) [inline cves/CVE-2015-1234.yml]: What was the weakness that it was hidden behind and why were they not testing it?

File: cves/CVE-2016-5149.yml [hunk 14]
```diff
@@ -50,28 +64,35 @@ reviews:
...
-- :commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
-  :note: ''
-- :commit: 79b64c3e741cc9c6afbb23885945831a45c6baa5
-  :note: ''
-- :commit: 4b8ca0a9fbbe79fa5a7819917e5a31d237f0f499
-  :note: ''
+- commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
+  note: 'This is the original commit that actually fixes the vulnerability.'
+- commit: 79b64c3e741cc9c6afbb23885945831a45c6baa5
+  note: 'This commit relanded the original fix and added another bug fix for CI/test failures.'
```
Comment (ConR24) [inline cves/CVE-2016-5149.yml]: I think you meant to use reverted rather than relanded in this line.

File: cves/CVE-2015-1234.yml [hunk 15]
```diff
@@ -47,24 +58,28 @@ reviews:
...
+    was particularly responsible for the CVE: GLsizei num_values = result->GetNumResults();
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes: 2
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2015-1234.yml]: Thanks!

File: cves/CVE-2015-1234.yml [hunk 16]
```diff
@@ -118,12 +137,23 @@ interesting_commits:
...
   commits:
-  - commit: 
-    note: 
-  - commit: 
-    note: 
+  - commit: d049874acef2be3c17612d4a06b480f3a45ea6e9
+    note: |
+      They were experiencing multiple minor bugs, so they preemptively added logging functionality
+      to multiple files, including the file that introduced the vulnerability. While this is a
+      safe and effective way to diagnose future issues, it seems like they were logging a bitt too
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2015-1234.yml]: Will push a fix. Thank you!

File: cves/CVE-2015-1234.yml [hunk 17]
```diff
@@ -132,7 +162,12 @@ major_events:
 
     The event doesn't need to be directly related to this vulnerability, rather,
     we want to capture what the development team was dealing with at the time.
-  answer: 
+  answer: |
+    No specific event was found, but it seems like the file in question was being worked on by
+    multiple people across some big companies (e.g. Intel, Samsung, Nvidia). Too many developers
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2015-1234.yml]: I don't think these people were hired to work on this project. It may have been a product of making this application open source.

File: cves/CVE-2015-1234.yml [hunk 18]
```diff
@@ -196,6 +240,27 @@ mistakes:
...
-  answer: 
+  answer: |
+    I feel that this vulnerability was more of a design mistake, followed by maintainability
+    issues. There is quite a bit of reliance on untrusted shared memory here because of the
+    way that the different processes communicate with each other, and this is part of the
+    architecture for this application, which means it is very difficult to change. This
+    reliance, combined with the complexity that comes naturally with concurrent processes,
+    exposes the application to a myriad of issues that may be very difficult to identify.
+
+    Moreover, there were multiple coding flaws that overlooked the atomicity of certain
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2015-1234.yml]: That's a good point, although I still feel that the way they designed the code to work lead to some oversights, which is the point I am trying to make here. I will try to make it clearer. Thank you!

File: cves/CVE-2016-5149.yml [hunk 19]
```diff
@@ -50,28 +64,35 @@ reviews:
...
   CVE-2011-3092.yml). Fixes and VCCs follow the same format.
 fixes:
-- :commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
-  :note: ''
-- :commit: 79b64c3e741cc9c6afbb23885945831a45c6baa5
-  :note: ''
-- :commit: 4b8ca0a9fbbe79fa5a7819917e5a31d237f0f499
-  :note: ''
+- commit: 91f655b19888da3f86b57ad8c548da93e7b9aba4
+  note: 'This is the original commit that actually fixes the vulnerability.'
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2016-5149.yml]: It involved computing the URL that they needed to use based on the security origin of the iframe. I will push a fix to make it a bit more clear. Thanks!

File: cves/CVE-2016-5149.yml [hunk 20]
```diff
@@ -50,28 +64,35 @@ reviews:
...
+    was possible in the first place.
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes: 2
```
[the remark this answers is not in the document]
Comment (gldias) [reply cves/CVE-2016-5149.yml]: Thank you!

File: cves/CVE-2016-5149.yml (hunk 12 above)
Comment (gldias) [reply cves/CVE-2016-5149.yml]: Thanks! I made some updates to it.

File: cves/CVE-2015-1234.yml (hunk 13 above)
Comment (gldias) [reply cves/CVE-2015-1234.yml]: The main issue was the possible remote execution of code, which was made possible by a race condition. The race condition was the weakness that was not being tested, which is understandable because concurrency always introduces unexpected issues. It is very difficult to test all use cases.

File: cves/CVE-2016-5149.yml (hunk 14 above)
Comment (gldias) [reply cves/CVE-2016-5149.yml]: Relanded means they re-introduced a commit that was previously reverted. This is the wording they used in the commit messages.

File: cves/CVE-2016-5149.yml (hunk 11 above)
Comment (gldias) [reply cves/CVE-2016-5149.yml]: I think that, in a way, it can be seen as a design mistake, but even with good design it is hard to predict all the possible race conditions in concurrent processes.
